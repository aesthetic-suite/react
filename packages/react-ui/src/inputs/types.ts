import { InputOptions } from 'reakit';
import { BorderSize, CommonSize, PaletteType } from '@aesthetic/react';
import { CommonProps } from '../types';
import { Shape } from '../types/shape';

export interface InputCommonProps extends InputOptions, CommonProps {
	/**
	 * Increase or decrease the border width.
	 * @default df
	 */
	border?: BorderSize;
	/**
	 * Customize the text, background, and border colors.
	 */
	palette?: Extract<PaletteType, 'negative' | 'positive' | 'primary' | 'warning'>;
	/**
	 * Customize the shape of the input (primarily border corner radius).
	 * @default round
	 */
	shape?: Shape;
	/**
	 * Increase or decrease the font size and spacing.
	 * @default df
	 */
	size?: CommonSize;
}
