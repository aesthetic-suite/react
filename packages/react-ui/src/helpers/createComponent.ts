/* eslint-disable @typescript-eslint/no-explicit-any */

import { forwardRef } from 'react';
import { CommonHtmlProps, HtmlElementType, OmitUnwantedHtmlProps } from '../types';

export function createComponent<Props extends object, Element extends HTMLElement = HTMLElement>(
	component: React.ForwardRefRenderFunction<Element, Props>,
) {
	return forwardRef(component);
}

export type DynamicProps<
	As extends React.ElementType,
	Props extends object
> = OmitUnwantedHtmlProps<React.ComponentPropsWithRef<As>> &
	Props & {
		/**
		 * Native HTML element or custom component to render as.
		 */
		as?: As;
	};

// This is based on the `React.ExoticComponent` but supports the dynamic `as` prop
// through a generic on the render function. We want the props on the consumer to change
// based on the `as` prop, but we *do not* want to type the props/ref in the implementation,
// as it makes the types too complex and we spread them into the DOM regardless.
export interface DynamicComponent<Props extends object, As extends React.ElementType> {
	<T extends As>(props: DynamicProps<T, Props>): React.ReactElement | null;
	readonly $$typeof: symbol;
	displayName?: string;
}

// We only want to return generic attributes when an HTML element,
// otherwise return the props from the custom component.
export type InferAsProps<T> = T extends HtmlElementType
	? CommonHtmlProps
	: T extends React.JSXElementConstructor<infer P>
	? P
	: {};

export function createDynamicComponent<
	Props extends object,
	As extends React.ElementType = HtmlElementType
>(component: React.ForwardRefRenderFunction<any, InferAsProps<As> & Props & { as?: As }>) {
	return (createComponent(component) as unknown) as DynamicComponent<Props, As>;
}
