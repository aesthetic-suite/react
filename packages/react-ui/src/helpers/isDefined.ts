export function isDefined(...values: unknown[]): boolean {
	return values.some((value) => typeof value !== 'undefined' && value !== null && value !== false);
}
