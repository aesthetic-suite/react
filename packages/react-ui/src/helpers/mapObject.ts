import { objectLoop } from '@aesthetic/utils';

export function mapObject<P, N>(
  base: Record<string, P>,
  onValue: (value: P) => N | undefined,
  onKey?: (key: string) => string,
): Record<string, N> {
  const object: Record<string, unknown> = {};

  objectLoop(base, (value, key) => {
    const nextValue = onValue(value);

    if (nextValue !== undefined) {
      object[onKey ? onKey(key) : key] = nextValue;
    }
  });

  return object as Record<string, N>;
}
