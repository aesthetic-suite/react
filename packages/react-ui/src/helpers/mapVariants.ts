import { ElementStyles } from '@aesthetic/react';
import { arrayLoop } from '@aesthetic/utils';

export function mapVariants<T>(
	type: string,
	enums: T[],
	factory: (value: T, typeName: string) => ElementStyles,
) {
	const object: Partial<Record<string, ElementStyles>> = {};

	arrayLoop(enums, (value) => {
		object[`${type}:${value}`] = factory(value, type);
	});

	return object as Record<string, ElementStyles>;
}
