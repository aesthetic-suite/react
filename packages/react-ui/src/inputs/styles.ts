import { PaletteType, style } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const styleSheet = style((css) => ({
	width: '100%',
	display: 'block',
	color: css.var('palette-neutral-text'),
	backgroundColor: css.var('palette-neutral-color-00'),
	borderColor: css.var(`palette-neutral-bg-base`),
	borderStyle: 'solid',
	appearance: 'none',
	outline: 'none',
	margin: 0,
	padding: 0,
	zIndex: 0,

	':disabled': {
		cursor: 'default',
	},

	'::placeholder': {
		color: css.var('palette-muted-text'),
	},

	'@variants': {
		...mapVariants(
			'palette',
			['negative', 'positive', 'primary', 'warning'] as PaletteType[],
			(palette) => ({
				borderColor: css.var(`palette-${palette}-bg-base`),

				':focus': { borderColor: css.var(`palette-${palette}-bg-focused`) },
				':hover': { borderColor: css.var(`palette-${palette}-bg-hovered`) },

				':disabled': {
					borderColor: css.var(`palette-${palette}-bg-disabled`),
					color: css.var(`palette-${palette}-fg-disabled`),
				},
			}),
		),
	},
}));
