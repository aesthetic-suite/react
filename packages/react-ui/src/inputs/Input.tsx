import React, { useMemo } from 'react';
import { Input as ReakitInput } from 'reakit';
import { useStyles } from '@aesthetic/react';
import { createComponent } from '../helpers/createComponent';
import { shapedBorderStyleSheet } from '../sheets/borders';
import { sizingStyleSheet } from '../sheets/sizing';
import { OmitUnwantedHtmlProps } from '../types';
import { styleSheet } from './styles';
import { InputCommonProps } from './types';

export interface InputProps
	extends InputCommonProps,
		OmitUnwantedHtmlProps<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {}

export const Input = createComponent<InputProps, HTMLInputElement>(function Input(
	{
		border = 'df',
		palette,
		shape = 'round',
		size = 'df',
		testID,
		// Inherited
		className: inheritedClassName,
		...props
	},
	ref,
) {
	const cx = useStyles(styleSheet, shapedBorderStyleSheet, sizingStyleSheet);
	const className = useMemo(
		() =>
			cx({ border, palette, shape, size }, 'element', 'borders', 'sizing', [inheritedClassName]),
		[cx, border, palette, shape, size, inheritedClassName],
	);

	return (
		<ReakitInput ref={ref} data-testid={testID} type="text" {...props} className={className} />
	);
});
