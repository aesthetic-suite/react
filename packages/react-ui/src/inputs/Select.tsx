import React, { useMemo } from 'react';
import { Input as ReakitInput } from 'reakit';
import { useStyles } from '@aesthetic/react';
import { createComponent } from '../helpers/createComponent';
import { shapedBorderStyleSheet } from '../sheets/borders';
import { sizingStyleSheet } from '../sheets/sizing';
import { OmitUnwantedHtmlProps } from '../types';
import { styleSheet } from './styles';
import { InputCommonProps } from './types';

export interface SelectProps
	extends InputCommonProps,
		OmitUnwantedHtmlProps<React.SelectHTMLAttributes<HTMLSelectElement>, 'size'> {
	/** List of `option` and `optgroup` elements. */
	children: NonNullable<React.ReactNode>;
}

export const Select = createComponent<SelectProps, HTMLSelectElement>(function Select(
	{
		border = 'df',
		children,
		palette,
		shape = 'round',
		size = 'df',
		testID,
		// Inherited
		className: inheritedClassName,
		...props
	},
	ref,
) {
	const cx = useStyles(styleSheet, shapedBorderStyleSheet, sizingStyleSheet);
	const className = useMemo(
		() =>
			cx({ border, palette, shape, size }, 'element', 'borders', 'sizing', [inheritedClassName]),
		[cx, border, palette, shape, size, inheritedClassName],
	);

	return (
		<ReakitInput ref={ref} as="select" data-testid={testID} {...props} className={className}>
			{children}
		</ReakitInput>
	);
});
