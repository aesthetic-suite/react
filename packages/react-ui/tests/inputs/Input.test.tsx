import React from 'react';
import { render, screen } from '@testing-library/react';
import { Input, InputProps } from '../../src/inputs/Input';
import { getRenderedStyles, withAccessibility, withStyles, Wrapper } from '../helpers';

function getElement() {
	return screen.getByPlaceholderText('Input')!;
}

describe('Input', () => {
	it('renders the input', () => {
		render(<Input placeholder="Input" />, { wrapper: Wrapper });

		expect(getElement().className).toBe(
			'element borders variant:border:df variant:shape:round variant:border:df sizing variant:size:df',
		);
	});

	it(
		'renders the input styles',
		withStyles(() => {
			render(<Input placeholder="Input" />, { wrapper: Wrapper });

			expect(getRenderedStyles('standard')).toMatchSnapshot();
		}),
	);

	it(
		'passes accessibility',
		withAccessibility(async (axe) => {
			const { container } = render(<Input placeholder="Input" />, { wrapper: Wrapper });

			expect(await axe(container)).toHaveNoViolations();
		}),
	);

	it('renders an `input` with type "text" by default', () => {
		render(<Input placeholder="Input" />, { wrapper: Wrapper });

		expect(getElement().tagName).toBe('INPUT');
		expect(getElement().getAttribute('type')).toBe('text');
	});

	it('can pass a test ID', () => {
		render(<Input placeholder="Input" testID="test" />, { wrapper: Wrapper });

		expect(getElement().getAttribute('data-testid')).toBe('test');
	});

	it('can pass a custom class name', () => {
		render(<Input className="foobar" placeholder="Input" />, { wrapper: Wrapper });

		expect(getElement().className).toEqual(expect.stringContaining('foobar'));
	});

	it('can pass native attributes', () => {
		render(<Input aria-label="Label" id="foo" placeholder="Input" type="email" />, {
			wrapper: Wrapper,
		});

		const el = getElement();

		expect(el.id).toBe('foo');
		expect(el.getAttribute('aria-label')).toBe('Label');
		expect(el.getAttribute('type')).toBe('email');
	});

	describe('props', () => {
		const props: Partial<InputProps> = {
			border: 'sm',
			palette: 'negative',
			shape: 'pill',
			size: 'lg',
		};

		Object.entries(props).forEach(([prop, value]) => {
			it(`sets "${prop}" class name`, () => {
				render(<Input placeholder="Input" {...{ [prop]: value }} />, { wrapper: Wrapper });

				expect(getElement().className).toEqual(expect.stringContaining(`variant:${prop}:${value}`));
			});
		});

		it('sets "type" attribute', () => {
			render(<Input placeholder="Input" type="password" />, { wrapper: Wrapper });

			expect(getElement().getAttribute('type')).toBe('password');
		});

		it('sets "value" attribute from "defaultValue"', () => {
			render(<Input defaultValue="foobar" placeholder="Input" />, { wrapper: Wrapper });

			expect((getElement() as HTMLInputElement).value).toBe('foobar');
		});

		it('sets "value" attribute from "value"', () => {
			render(<Input placeholder="Input" value="foobar" onChange={jest.fn()} />, {
				wrapper: Wrapper,
			});

			expect((getElement() as HTMLInputElement).value).toBe('foobar');
		});
	});
});
