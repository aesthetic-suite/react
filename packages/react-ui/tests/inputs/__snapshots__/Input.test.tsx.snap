// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders the input styles 1`] = `".ctfda3b {width: 100%;}.c1s7hmty {display: block;}.c1u0bpqh {color: var(--palette-neutral-text);}.cwk51if {background-color: var(--palette-neutral-color-00);}.cfrhdt4 {border-color: var(--palette-neutral-bg-base);}.cfdmyha {border-style: solid;}.c16lb9qp {appearance: none;}.cxfr5f3 {outline: none;}.c13kbekr {margin: 0;}.cqmlbj8 {padding: 0;}.c2uri1o {z-index: 0;}.cnv1cj2:disabled {cursor: default;}.c1oxedqq::placeholder {color: var(--palette-muted-text);}.c1kgx3tq {border-color: var(--palette-negative-bg-base);}.c91s500:focus {border-color: var(--palette-negative-bg-focused);}.c2c8ej4:hover {border-color: var(--palette-negative-bg-hovered);}.c1ljzbl:disabled {border-color: var(--palette-negative-bg-disabled);}.cxxlv8:disabled {color: var(--palette-negative-fg-disabled);}.c1syia52 {border-color: var(--palette-positive-bg-base);}.c1lipdso:focus {border-color: var(--palette-positive-bg-focused);}.c1tga120:hover {border-color: var(--palette-positive-bg-hovered);}.c1857xzt:disabled {border-color: var(--palette-positive-bg-disabled);}.c16kxa2k:disabled {color: var(--palette-positive-fg-disabled);}.c179czo1 {border-color: var(--palette-primary-bg-base);}.cg3llcv:focus {border-color: var(--palette-primary-bg-focused);}.c1fcxwbj:hover {border-color: var(--palette-primary-bg-hovered);}.c12f7n4u:disabled {border-color: var(--palette-primary-bg-disabled);}.cua9xzv:disabled {color: var(--palette-primary-fg-disabled);}.cin4dcn {border-color: var(--palette-warning-bg-base);}.c1woskih:focus {border-color: var(--palette-warning-bg-focused);}.cdsjra1:hover {border-color: var(--palette-warning-bg-hovered);}.cfjma3c:disabled {border-color: var(--palette-warning-bg-disabled);}.co7eq8t:disabled {color: var(--palette-warning-fg-disabled);}.c1e95qeq {border-width: var(--border-sm-width);}.c1q9bcr2 {border-width: var(--border-df-width);}.c1xgoolj {border-width: var(--border-lg-width);}.c18kfffx {border-radius: 100px;}.c18pn6tu {border-radius: var(--border-sm-radius);}.c1rooa26 {border-radius: var(--border-df-radius);}.cmt9y9z {border-radius: var(--border-lg-radius);}.c433b9o {border-radius: 0;}.cxzk1cy {line-height: var(--text-sm-line-height);}.c1qx7d18 {font-size: var(--text-sm-size);}.cfnwgvv {padding: 0.31rem 0.94rem;}.c1ts5a2r {min-width: 70px;}.c1q1da5a {line-height: var(--text-df-line-height);}.cpfxegw {font-size: var(--text-df-size);}.c1rmjobf {padding: 0.94rem 1.56rem;}.cjjixdc {min-width: 87.5px;}.c10l4oaf {line-height: var(--text-lg-line-height);}.ckfuq95 {font-size: var(--text-lg-size);}.c14wns5t {padding: 1.56rem 2.50rem;}.cwb63sg {min-width: 105px;}"`;
